packages/rich-text/src/Toolbar/index.tsx

```diff
@@ -42,61 +45,142 @@ const styles = {
   const onCloseEntityDropdown = () => setEmbedDropdownOpen(false);
   const onToggleEntityDropdown = () => setEmbedDropdownOpen(!isEmbedDropdownOpen);
 
+  const [canAccessAssets, setCanAccessAssets] = useState(false);
+  React.useEffect(() => {
+    sdk.access.can('read', 'Asset').then(setCanAccessAssets);
+  }, [sdk]);
+  const inlineEntryEmbedEnabled = isNodeTypeEnabled(sdk.field, INLINES.EMBEDDED_ENTRY);
+  const blockEntryEmbedEnabled = isNodeTypeEnabled(sdk.field, BLOCKS.EMBEDDED_ENTRY);
+  const blockAssetEmbedEnabled =
+    canAccessAssets && isNodeTypeEnabled(sdk.field, BLOCKS.EMBEDDED_ASSET);
+
+  const numEnabledEmbeds = [
+    inlineEntryEmbedEnabled,
+    blockEntryEmbedEnabled,
+    blockAssetEmbedEnabled,
+  ].filter(Boolean).length;
+  const shouldDisplayDropdown = numEnabledEmbeds > 1;
+
+  const icons = (
+    <Fragment>
+      {isNodeTypeEnabled(sdk.field, BLOCKS.EMBEDDED_ENTRY) && (
+        <EmbeddedEntityBlockToolbarIcon
+          isDisabled={!!isDisabled}
+          nodeType={BLOCKS.EMBEDDED_ENTRY}
+          onClose={onCloseEntityDropdown}
+          isButton={!shouldDisplayDropdown}
+        />
+      )}
+      {isNodeTypeEnabled(sdk.field, INLINES.EMBEDDED_ENTRY) && (
+        <ToolbarEmbeddedEntityInlineButton
+          isDisabled={!!isDisabled}
+          onClose={onCloseEntityDropdown}
+        />
+      )}
+      {isNodeTypeEnabled(sdk.field, BLOCKS.EMBEDDED_ASSET) && (
+        <EmbeddedEntityBlockToolbarIcon
+          isDisabled={!!isDisabled}
+          nodeType={BLOCKS.EMBEDDED_ASSET}
+          onClose={onCloseEntityDropdown}
+        />
+      )}
+    </Fragment>
+  );
+
+  if (shouldDisplayDropdown) {
+    return (
+      <EmbeddedEntityDropdownButton
+        isDisabled={isDisabled}
+        onClose={onCloseEntityDropdown}
+        onToggle={onToggleEntityDropdown}
+        isOpen={isEmbedDropdownOpen}>
+        {icons}
+      </EmbeddedEntityDropdownButton>
+    );
+  } else {
+    return icons;
+  }
+};
+
+const Toolbar = ({ isDisabled }: ToolbarProps) => {
+  const sdk = useSdkContext();
+  const editor = useStoreEditor() as SPEditor;
+  const canInsertBlocks = !isNodeTypeSelected(editor, BLOCKS.TABLE);
+  const validationInfo = React.useMemo(() => getValidationInfo(sdk.field), [sdk.field]);
+
   return (
     <EditorToolbar testId="toolbar">
       <div className={styles.formattingOptionsWrapper}>
         <ToolbarHeadingButton isDisabled={isDisabled || !canInsertBlocks} />
 
-        <EditorToolbarDivider />
-
-        <ToolbarBoldButton isDisabled={isDisabled} />
-        <ToolbarItalicButton isDisabled={isDisabled} />
-        <ToolbarUnderlineButton isDisabled={isDisabled} />
-        <ToolbarCodeButton isDisabled={isDisabled} />
+        {validationInfo.isAnyMarkEnabled && <EditorToolbarDivider />}
 
-        <EditorToolbarDivider />
+        {isMarkEnabled(sdk.field, MARKS.BOLD) && <ToolbarBoldButton isDisabled={isDisabled} />}
```

Comment:
What's the difference between being `enabled` and `disabled` here? Can it be enabled and disabled at the same time? 😆 

I'm asking that because the `isDisabled` prop seems not necessary.

Reply:
The `isDisabled` prop represents the button as disabled when the editor as a whole is in a disabled state, for example in the UI's read-only version comparison (diff) mode, or when an entry- or field-level error prevents the editor from being accessible. In cases like these, we still want buttons for valid marks and nodes to be represented, but displayed as disabled. By contrast, when a mark or node is disabled by the content model, we don't want to display the buttons to begin with.